tests/test_array_namespace.py

```diff
@@ -26,6 +26,22 @@ def test_array_namespace_multiple():
     assert array_namespace(x, x) == array_namespace((x, x)) == \
         array_namespace((x, x), x) == array_api_compat.numpy
 
+def test_fallback_namespace():
+    import numpy as np
+    import array_api_compat.numpy
+
+    xp = array_api_compat.numpy
+    xp_ = array_namespace([1, 2], fallback_namespace=xp)
+    assert xp_ == xp
+
+    xp_ = array_namespace([1, 2], np.asarray([1, 2]), fallback_namespace=xp)
```

Comment:
Can we add a test to make sure the fallback namespace only applies when it's ambiguous. Something like `array_namespace([1, 2], np.asarray([1, 2]), fallback_namespace=numpy.array_api)` should return `array_api_compat.numpy`.

Reply:
I am not sure how this would work. To me that's the purpose of the fallback. With

`array_namespace([1, 2], np.asarray([1, 2]), fallback_namespace=numpy.array_api)`

I would expect `[1, 2]` to fallback to `numpy.array_api`. I suppose that in this case we could say ok fallback `[1, 2]` to be the same as the other namespace. But what if instead we have:

`array_namespace([1, 2], cp.asarray([1, 2]), fallback_namespace=numpy.array_api)`

Shall this also simplify to `cp`??

(Due to the recursions it's also getting tricky to follow the logic.)

Reply:
Maybe I'm missing the point of this then. In general I'd say you shouldn't mix different array libraries with the array API. Some libraries might support it (like NumPy being able to convert other arrays), and in some cases with a performance penalty (like moving from GPU to CPU). So `array_namespace([1, 2], cp.asarray([1, 2]), fallback_namespace=numpy.array_api)` should give an error. 

So to me the only point of this is to allow this function to also accept non-array inputs, like Python scalars and/or lists of Python scalars. You can convert those to arrays with `asarray`, but you can't call `asarray` until you have a namespace, making it a chicken/egg situation. Of course, then it might still be a good idea to make sure the errors from this function are helpful. 

I also tend to agree that the best way to handle this is to simply not allow non-array inputs to functions that use the array API. That forces the users to resolve the ambiguity of what array library the want to use by calling asarray themselves.

Reply:
By the way, this function automatically denests lists of arrays. This was to internally support functions like `concat` which accept lists of arrays. But I'm wondering if this behavior should be removed, so that there there is less ambiguity and we can unconditionally error on `list` or `tuple` input.

Reply:
Yeah the main purpose of the proposal was to assume a namespace for things that would otherwise be marked as non compatible or unknown (list, scalar, tuples, else). Then yes it's on the user to call `asarray` and whether or not it works is on their side.

> By the way, this function automatically denests lists of arrays.

Yeah this is making the logic difficult IMHO. I was having trouble to find a way to address your comment with that in place. If that can be removed, I am happy to do so (here or another PR, let me know.)

Reply:
> Yeah the main purpose of the proposal was to assume a namespace for things that would otherwise be marked as non compatible or unknown (list, scalar, tuples, else).

I think that should live in SciPy, because it's orthogonal to array API support. Use of lists and other "array-likes" is specific to NumPy, and is by and large considered a design mistake. The array API standard does not allow array-likes, and neither do other libraries like PyTorch and CuPy. So there's no reason for this package to care about array-likes.

Reply:
> If that can be removed, I am happy to do so (here or another PR, let me know.)

If you can do that it would be great. I'm still unclear what the changes here are (do we actually want this or not?), so maybe a separate PR is appropriate. You'll need to fix the functions like `concat` that do accept a list. Right now the wrappers assume they can just pass things through to array_namespace.

Reply:
> I think that should live in SciPy, because it's orthogonal to array API support. Use of lists and other "array-likes" is specific to NumPy, and is by and large considered a design mistake. The array API standard does not allow array-likes, and neither do other libraries like PyTorch and CuPy. So there's no reason for this package to care about array-likes.

I agree. For functions that are wrapped functions from NumPy (or any other library), we should support whatever the wrapped function supports to maintain maximal compatibility. But for new functions, we can be a little more strict. And generally speaking `array_namespace()` is an important function that will be used everywhere so we should try to discourage bad habits with it.

Reply:
Ok then it looks like we could close this PR and I could try to propose a PR to only accept Array API compatible arrays. Is that correct?

Reply:
Yes that sounds good.